src/Raven.Server/Documents/RetireAttachmentsSender.cs

```diff
@@ -115,32 +107,66 @@ internal async Task<int> RetireAttachments(int batchSize, long maxItemsToProcess
-                                var collection = doc.Id;
-
                                 if (CanContinueBatch(Logger, duration, totalUploaded) == false)
                                 {
                                     break;
                                 }
 
-                                if (string.IsNullOrEmpty(collection))
+                                if (string.IsNullOrEmpty(doc.Id))
                                 {
                                     if (Logger.IsInfoEnabled)
-                                        Logger.Info($"Skipping '{nameof(AttachmentRetireType.PutRetire)}' of retired attachment with key: '{key}' because it's collection '{collection}' IsNullOrEmpty.");
+                                        Logger.Info($"Skipping PutRetire of retired attachment with key: '{doc.LowerId}' because it's identifier '{doc.Id}' IsNullOrEmpty.");
 
                                     // document was deleted, need to remove it from retired tree
                                     retired.Enqueue(doc);
                                     continue;
                                 }
 
-                                var hash = _database.DocumentsStorage.AttachmentsStorage.ExtractHashFromAttachmentId(doc.LowerId);
+
+                                // get uploader for the attachment
+                                if (directUploaders.TryGetValue(doc.Id, out var directUpload) == false)
+                                {
+                                    // check if we have a destination configuration for the identifier
+
+                                    if (Configuration.Destinations.TryGetValue(doc.Id, out var destination) == false)
+                                    {
+                                        // no destination found for the identifier
+                                        if (Logger.IsWarnEnabled)
+                                            Logger.Warn($"No destination found for retired attachment with key: '{doc.LowerId}' and identifier '{doc.Id}'. Will try to retire it again!");
+
+                                        // this will keep the retired attachment in the tree and will try to retire it again afterwards.
+
+                                        totalCount--; // let's decrease the total count in case we have a lot of attachments without destination that we skip
+                                        continue;
+                                    }
+
+                                    if (destination.HasUploader() == false)
+                                    {
+                                        // no uploader for this destination
+                                        if (Logger.IsWarnEnabled)
+                                            Logger.Warn($"No uploader found for destination with identifier '{destination.Identifier}' for attachment with key: '{doc.LowerId}'. Will try to retire it again!");
```

Comment:
what does this mean? There is a destination but no uploader?

Reply:
it means there is a destination but it doesn't have S3 or Azure settings:
https://github.com/ravendb/ravendb/blob/4119158f181d9a61c13c53d383e0a0f891aacbab/src/Raven.Client/Documents/Attachments/RetiredAttachmentsDestinationConfiguration.cs#L83

It should not be possible by adding the configuration by operation, since we do AssertConfiguration here:
 https://github.com/ravendb/ravendb/blob/4119158f181d9a61c13c53d383e0a0f891aacbab/src/Raven.Client/Documents/Attachments/RetiredAttachmentsConfiguration.cs#L96
 
 But it might happen when DatabaseRecord is edited manually, so I added this check.